### PR TITLE
Add config option to forget rooms automatically when users leave them

### PR DESCRIPTION
**Pull request by squahtx**
This is largely based off the stats and user directory updater code.

---

Should be reviewable commit by commit. I can split this up into multiple PRs if needed.

### CODE REVIEW

File: synapse/handlers/room_member.py [hunk 1]
```diff
@@ -278,9 +281,25 @@ async def _user_left_room(self, target: UserID, room_id: str) -> None:
...
+            Membership.LEAVE,
+            Membership.BAN,
+        ]:
+            raise SynapseError(400, "User %s in room %s" % (user_id, room_id))
+
+        # In normal case this call is only required if `membership` is not `None`.
+        # But: After the last member had left the room, the background update
+        # `_background_remove_left_rooms` is deleting rows related to this room from
+        # the table `current_state_events` and `get_current_state_events` is `None`.
+        await self.store.forget(user_id, room_id)
```
Comment (squahtx) [inline synapse/handlers/room_member.py:302]: I couldn't see why forgetting rooms couldn't be done on any worker, so I made the action available to all workers.
There may be something I'm missing?

File: synapse/handlers/room_member.py [hunk 2]
```diff
@@ -2018,25 +2037,139 @@ async def _user_left_room(self, target: UserID, room_id: str) -> None:
...
-            raise SynapseError(400, "User %s in room %s" % (user_id, room_id))
+    For the purposes of this feature, kicks, bans and "leaves" via state resolution
+    weirdness are all considered to be leaves.
 
-        # In normal case this call is only required if `membership` is not `None`.
-        # But: After the last member had left the room, the background update
-        # `_background_remove_left_rooms` is deleting rows related to this room from
-        # the table `current_state_events` and `get_current_state_events` is `None`.
-        await self.store.forget(user_id, room_id)
+    Derived from `StatsHandler` and `UserDirectoryHandler`.
```
Comment (squahtx) [inline synapse/handlers/room_member.py:2047]: I wonder if it's worth factoring out the duplicated structure. There is certainly a lot of faff involved in adding another one of these state delta-consumers.

File: synapse/storage/databases/main/roommember.py [hunk 3]
```diff
@@ -1368,6 +1368,55 @@ def _is_local_host_in_room_ignoring_users_txn(
...
+                "UPDATE"
+                "  room_memberships"
+                " SET"
+                "  forgotten = 1"
+                " WHERE"
+                "  user_id = ?"
+                " AND"
+                "  room_id = ?"
+            )
+            txn.execute(sql, (user_id, room_id))
```
Comment (erikjohnston) [inline synapse/storage/databases/main/roommember.py:1385]: While we're here can we change this to a `simple_update` please? The query hurts my eyes.

File: synapse/handlers/room_member.py [hunk 4]
```diff
@@ -2018,25 +2037,139 @@ async def _user_left_room(self, target: UserID, room_id: str) -> None:
...
+                    self.pos, room_max_stream_ordering
+                )
+
+                logger.debug("Handling %d state deltas", len(deltas))
+                await self._handle_deltas(deltas)
+            else:
+                # Update the processing position, so that if the server admin turns the
+                # feature on at a later date, we don't decide to forget every room that
+                # has ever been left in the past.
+                pass
```
Comment (erikjohnston) [inline synapse/handlers/room_member.py:2130]: Can we lift this out of the loop and instead have a top level:

```python
if not self._hs.config.room.forget_on_leave:
    await self._store.update_room_forgetter_stream_pos(max_pos)
    return
```

I think that's clearer?

Comment (erikjohnston) [reply synapse/handlers/room_member.py:2130]: Also, `max_pos` isn't defined in this branch

File: changelog.d/15224.feature [hunk 5]
```diff
@@ -0,0 +1 @@
+Add `forget_rooms_on_leave` config option to automatically forget rooms when users leave them or are removed from them.
```
Comment (MadLittleMods) [inline changelog.d/15224.feature:1]: Does this fix https://github.com/matrix-org/synapse/issues/4720 ?

We should update the issue description in that case

Comment (squahtx) [reply changelog.d/15224.feature:1]: No, this does not purge rooms from the database at all.

Comment (squahtx) [reply changelog.d/15224.feature:1]: This PR is adjacent, but I see it as more optional than anything. #4720 sounds like it's scoped to the case where all users explicitly choose to forget a room themselves. This PR forces users to forget rooms, making the proposed GC in #4720 trigger more often.

Comment (MadLittleMods) [reply changelog.d/15224.feature:1]: 🤷 Feels like a nice step to make the GC automatic (sane default)

File: synapse/handlers/room_member.py (hunk 4 above)
Comment (squahtx) [reply synapse/handlers/room_member.py]: done

**Review by erikjohnston — APPROVED** [review]

**Pull request merged**